### PR TITLE
LSP: Implement semantic token requests

### PR DESCRIPTION
**Pull request by Marenz**

### CODE REVIEW

File: libsolidity/lsp/SemanticTokensBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+#include <liblangutil/SourceLocation.h>
+
+#include <fmt/format.h>
+
+using namespace std;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+#if 0 // !defined(NDEBUG)
+#define DPRINT(x) do { std::cerr << (x) << std::endl; } while (0)
```
Comment (christianparpart) [inline libsolidity/lsp/SemanticTokensBuilder.cpp]: This is old take-over code that can go now. Use lspDebug(x) for consistency instead. :)

File: libsolidity/lsp/SemanticTokensBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+	case frontend::Type::Category::Bool: return SemanticTokenType::Number;
+	case frontend::Type::Category::Enum: return SemanticTokenType::Enum;
+	case frontend::Type::Category::Function: return SemanticTokenType::Function;
+	case frontend::Type::Category::Integer: return SemanticTokenType::Number;
+	case frontend::Type::Category::RationalNumber: return SemanticTokenType::Number;
+	case frontend::Type::Category::StringLiteral: return SemanticTokenType::String;
+	case frontend::Type::Category::Struct: return SemanticTokenType::Struct;
+	case frontend::Type::Category::Contract: return SemanticTokenType::Class;
+	default:
+		DPRINT(fmt::format("semanticTokenTypeForType: unknown category: {}", static_cast<unsigned>(_type->category())));
```
Comment (christianparpart) [inline libsolidity/lsp/SemanticTokensBuilder.cpp]: I think we could be even doing `solAssert(false, "")` here.

File: libsolidity/lsp/SemanticTokensBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+		encode(rhsLocation, memberTokenType);
+	}
+
+	return false; // we handle LHS and RHS explicitly above.
+}
+
+bool SemanticTokensBuilder::visit(frontend::ParameterList const& _node)
+{
+	return true;
+	(void) _node;
```
Comment (christianparpart) [inline libsolidity/lsp/SemanticTokensBuilder.cpp]: The whole function can be removed.

File: libsolidity/lsp/LanguageServer.cpp [hunk 4]
```diff
@@ -266,10 +311,31 @@ void LanguageServer::handleInitialize(MessageID _id, Json::Value const& _args)
...
+{
+	auto uri = _args["textDocument"]["uri"];
+
+	compile();
+
+	auto const sourceName = m_fileRepository.uriToSourceUnitName(uri.as<string>());
+	SourceUnit const& ast = m_compilerStack.ast(sourceName);
+	m_compilerStack.charStream(sourceName);
+	SemanticTokensBuilder semanticTokensBuilder;
+	Json::Value data = semanticTokensBuilder.build(ast, m_compilerStack.charStream(sourceName));
```
Comment (christianparpart) [inline libsolidity/lsp/LanguageServer.cpp]:
```suggestion
	Json::Value data = SemanticTokensBuilder().build(ast, m_compilerStack.charStream(sourceName));
```
Shouldn't that do it, too?

**Review by christianparpart — CHANGES_REQUESTED** [review]

File: libsolidity/lsp/SemanticTokensBuilder.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/lsp/SemanticTokensBuilder.cpp]: Then we would throw for arrays

**Review by christianparpart — APPROVED** [review]
LGTM after interactive review. :)

File: Changelog.md [hunk 5]
```diff
@@ -4,6 +4,7 @@ Language Features:
 
 
 Compiler Features:
+* LSP: Add rudimentary support for semantic highlighting.
```
Comment (ekpyron) [inline Changelog.md:7]:
```suggestion
 * LSP: Add rudimentary support for semantic highlighting.
```
... :-)

**Pull request merged**